### ISSUE TITLE
[HOLD for payment 2024-05-30] [$500] Desktop - Chat isn't marked as read if user clicks on notification from Notification Center

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#34829]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.28-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Open Desktop App
2. Login as User A
3. From the second device with account of User B sent a message to User A
4. User A received Notification
5. Wait until Notification goes to the Notification center
6. Open Notification center and click on this received notification

## Expected Result:
User goes to the correct chat, new messages are marked as read and this chat is no longer marked in bold

## Actual Result:
Chat is marked in bold
=====
If User clicks on a notification BEFORE it goes to the notification center, the chat will be marked as read.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/aff3d35b-99ca-4985-a32b-ed1885ec91fc

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011a42b3684380c039</li>
        <li>Upwork Job ID: 1748411136068030464</li>
        <li>Last Price Increase: 2024-01-19</li>
<li>Automatic offers: </li>
<ul>
<li>ishpaul777 | Reviewer | 28118184</li>
<li>wildan-m | Contributor | 0</li>
</ul></li>
    </ul>
</details>

**Comment by mirimichaelson** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

We want to make sure the chat is marked as read when messages have gone into the Notification Center and are clicked from there.

### What is the root cause of that problem?

The root of the problem is that the App is not in the ‘Visible’ state when the Notification Center message is clicked. There’s a small if statement in `ReportActionsList.js` which ensures that if the app isn’t visible (Visibility.isVisible()), messages will not be read.  

```jsx
if (Visibility.isVisible() && scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD) {
     Report.readNewestAction(report.reportID);
} else {
     readActionSkipped.current = true;
}
```

This is useful in some cases, like receiving messages if you’re afk, but unhelpful in multiple other cases where the same bug shows up; 

1) You click on a notification within the Notification Center

2) Click on a notification before it goes into the Notification Center, if you have another tab open in front of Expensify (same bug as this ticket)

3) You’re already chatting to someone and receive new messages. The bolding of the text doesn’t go away, until you click to another chat and back again. So again, the same bug. (See the attached video)

### What changes do you think we should make in order to solve the problem?

This is the first bug I’m looking into, so I don’t know exactly what your desired behaviour is.

My suggestion would be adding to the functionality that makes chats visible:

Right now, the functionality for updating to the ‘Read’ state requires clicking on a message in the sidebar. I would say that we should add additional functionality so that if a user clicks within one of their chats, or on the chat title in the left hand sidebar, it will trigger the useEffect to update, and mark messages as read. This will resolve all three bugs that currently exist.

### What alternative solutions did you explore? (Optional)

The simple code fix is removing `isVisible` , but this is at the expense of causing other issues - so messages will mark as read, even if the user is viewing other apps in front of it.

[Screen Recording 2024-01-20 at 04.10.11.mov.zip](https://github.com/Expensify/App/files/13996675/Screen.Recording.2024-01-20.at.04.10.11.mov.zip)

**Comment by mirimichaelson** [comment]
Contributor details
Your Expensify account email: miriammichaelson@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01cce98dffb320182a

**Comment by JeremyCroff** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The unread status styling of a report does not disappear when clicking on the OS notification when it's docked in the Notification Center.

### What is the root cause of that problem?
At https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsList.js#L211 The utility that exposes the Visibility has a stale reference to the document state.

### What changes do you think we should make in order to solve the problem?
If you replace the Visibility.isVisible() with the actual document.visibilityState === 'visible', it will behave as expected. 



### What alternative solutions did you explore? (Optional)
I explored removing the visibility, but this impact the user experience. Also looked at other window methods such as hidden, and focus.

**Comment by garrettmknight** [comment]
@sobitneupane what do you think of this first proposal?

**Comment by jeremy-croff** [comment]
My proposal on this is temporarily hidden because my primary github account is suspended while I am recovering it.
My solution was that I identified the existing 
Visibility.isVisible() check gets a stale reference of the notification app.
Calling "document.visibilityState === 'visible'" directly in the ReportActionsList getse a fresh window state and clears the chat reliably.

Do I need to write a new proposal for this? It should be in the history of emails/notifications

**Comment by jeremy-croff** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The Chat does not get marked as read if the OS notification gets clicked from the docked notification center state.

### What is the root cause of that problem?
Visibility.isVisible() utility function unreliably returns false when the notification is from the docked notification center. This seems to be a bug with the util function, and the window document state from electron.

### What changes do you think we should make in order to solve the problem?
When replacing Visibility.isVisible() in the ReportActionsList ( where it checks if it's visible before it marks as read) 
with the actual document.visibilityState === 'visible', it works as expected. The utility has an unreliable reference in this specific scenario

**Comment by sobitneupane** [comment]
I am not around and won't be able to review it soon. @garrettmknight Can we get someone else to review proposal?

**Comment by garrettmknight** [comment]
Yeah, no problem.

**Comment by garrettmknight** [comment]
Hey @ishpaul777 - we've got a couple proposals for this one when you've got a minute to review!

**Comment by jeremy-croff** [comment]
> Thanks for your proposal @jeremy-croff, RCA makes and solution tests well, Can you do a impact analysis of your solution just want to make sure we didn't break current usage of`Visibility.isVisible`.

- Identified that the original visibility to focus mapped logic was implemented when electron had no support for visibility yet. https://github.com/Expensify/App/commit/e7150f92bfd50e5d46ba65e85163bdac808a2229
This gives good indication is was a tech limitation rather than an intentional choice.

- When reviewing out uses of isVisibility, visibility is a more optimistic truthy value than isFocused, so I am reviewing it with the lens of are any unintentional behaviours happening that shouldn't. 

- When reviewing impact analysis, the change would be scoped to changes in the electron main file's interpration of visibility which can be found here https://github.com/Expensify/App/blob/main/desktop/main.js#L558, so it wouldn't impact other platforms.

Line by line analysis:
https://github.com/Expensify/App/blob/5edf172457de71f5fa77d9ddf30e03726db346a2/src/pages/signin/LoginForm/BaseLoginForm.js#L280 will behave the same because of the last condition will be truthy always

https://github.com/Expensify/App/blob/c3f7bccc1050988c6352ea63fa00bf327761587a/src/libs/actions/Report.ts#L1834 behaves the same because of the last condition will be truthy always

https://github.com/Expensify/App/blob/c3f7bccc1050988c6352ea63fa00bf327761587a/src/libs/getIsReportFullyVisible.ts#L6 will behave different, but we can implement isFocused seperate for electron ( currently short circuits to true ) to achieve the same logical behaviour here https://github.com/Expensify/App/blob/c3f7bccc1050988c6352ea63fa00bf327761587a/src/libs/Visibility/index.desktop.ts#L11, and then a listener on the main process again.

https://github.com/Expensify/App/blob/c3f7bccc1050988c6352ea63fa00bf327761587a/src/components/Form/FormProvider.js#L324 will benefit from similarly from also having an seperate focus implementation


In short, having a seperate isFocus implementation will further align the behavior between web and focus, and mitigate any impact.

**Comment by jeremy-croff** [comment]
> Thanks for a detailed analysis. @jeremy-croff
> 
> > but we can implement isFocused seperate for electron (currently short circuits to true ) to achieve the same logical behaviour here
> 
> Can you update your solution to include the changes required for the other cases to work as expected

Updated

**Comment by blimpich** [comment]
Sounds good. @jeremy-croff feel free to put up a PR when you have it ready.

@mirimichaelson thank you for posting your first proposal! It looks like you put a lot of effort into figuring out a possible solution, and I appreciate the detail. Although your proposal wasn't chosen I'd encourage you to keep looking around for possible issues to pick up 🙂. Thank you for contributing to this issue's discussion and welcome to the repository!

**Comment by blimpich** [comment]
Update: we're waiting to see if we can get an Electron maintainer to complete an upstream issue that will impact this issue. [Comment here](https://github.com/electron/electron/issues/38955#issuecomment-1908825876). If no word comes from them then we'll look into this again sometime next week.

**Comment by garrettmknight** [comment]
Sounds good - will leave this daily for checkins.

**Comment by blimpich** [comment]
We have [confirmation](https://github.com/electron/electron/issues/38955#issuecomment-1921147336) from the Electron dev that they are working on the upstream issue. They said they hope to get to it next week. @ishpaul777 @jeremy-croff.

**Comment by jeremy-croff** [comment]
> We have [confirmation](https://github.com/electron/electron/issues/38955#issuecomment-1921147336) from the Electron dev that they are working on the upstream issue. They said they hope to get to it next week.
> 
> cc: @ishpaul777 @jeremy-croff.

Okay great.
So what we can do for this PR is follow the original proposal, and just implement the correct electron visibility method. End of PR. Simple.

I had referenced some other github issues that the onfocus listener implementation would fix, but i'll pull it out comment on those issues with the proposal seperate.

**Comment by blimpich** [comment]
@ishpaul777 Good point! Let me see if I can get the ball rolling on getting our electron version upgraded while we wait for this upstream patch to happen. 

> what should be process for that do we need consensus for upgrade?

Here's my plan:
1. create a separate issue for bumping our electron version 28. I think this will be an easy sell since upgrading a core library that we use is important for staying up to date
2. Once the electron library is upgraded and the upstream issue that effects this issue is patched, then we can finish this issue. Per [Jeremy's comment](https://github.com/Expensify/App/issues/34829#issuecomment-1922127275) I think there is still work to be done here irrespective of the electron version bump correct?

If that sounds good to both of you I think we can change the priority of this to weekly. Does that sound like a plan?

**Comment by jeremy-croff** [comment]
> @ishpaul777 Good point! Let me see if I can get the ball rolling on getting our electron version upgraded while we wait for this upstream patch to happen.
> 
> > what should be process for that do we need consensus for upgrade?
> 
> Here's my plan:
> 
> 1. create a separate issue for bumping our electron version 28. I think this will be an easy sell since upgrading a core library that we use is important for staying up to date
> 2. Once the electron library is upgraded and the upstream issue that effects this issue is patched, then we can finish this issue. Per [Jeremy's comment](https://github.com/Expensify/App/issues/34829#issuecomment-1922127275) I think there is still work to be done here irrespective of the electron version bump correct?
> 
> If that sounds good to both of you I think we can change the priority of this to weekly. Does that sound like a plan?
Seems we can put the PR on hold until after we have a fix version in electron, 
and completed the https://github.com/Expensify/App/issues/35600 upgrade Issue.
If they are targeting feb 20 this probably will be on hold for a good month or so.

**Comment by blimpich** [comment]
Upstream issue is officially fixed 🎉 https://github.com/electron/electron/issues/38955#issuecomment-1929219329

**Comment by blimpich** [comment]
It looks like this issue is going to be patched in electron 25, so we might not even need to upgrade major versions in order to make use of the fix. It got merged to main but looks like we are waiting for it to be patched specifically in version 25 https://github.com/electron/electron/pull/38982#issuecomment-1929218085.

I'll keep the major version upgrade issue open though since that is worth doing in its own right.

**Comment by FitseTLT** [comment]
I think it would be better to hold this for https://github.com/Expensify/App/issues/33680

**Comment by jeremy-croff** [comment]
@ishpaul777 He is implementing the focus listener logic. Which I had identified as solving the electron issue, as well as the LHN bold issues for desktop. A previous few commits of mine was doing this in the same way.

But it was considered a workaround for the electron bug, with additional risk, and I was advised on the PR to implement the minimum to resolve the bug by implementing the electron visibility apis. I think we will benefit from having the correct electron API's implemented. But if his PR goes first, I think his logic will kick in and make it non reproducable. I can test his branch later?

**Comment by garrettmknight** [comment]
@jeremy-croff you get a chance to test his branch and see how it worked?

**Comment by jeremy-croff** [comment]
> @jeremy-croff you get a chance to test his branch and see how it worked?

Bookmarked to test the branch in approx 5 hours. Confidence is pretty high it will be addressed

**Comment by jeremy-croff** [comment]
> @jeremy-croff you get a chance to test his branch and see how it worked?

Confirmed this addresses this bug

**Comment by garrettmknight** [comment]
Updating this to HOLD in that case for https://github.com/Expensify/App/issues/33680 and https://github.com/electron/electron/pull/41314 - will continue to check on those before retesting/taking his off hold if the bug still exists.

**Comment by garrettmknight** [comment]
Just a note this is still on hold - keeping it while I go OOO because I doubt 33680 goes to prod till I get back.

**Comment by garrettmknight** [comment]
Still on hold

**Comment by garrettmknight** [comment]
Still on hold for both issues

**Comment by jeremy-croff** [comment]
@ishpaul777 I have confirmed that this is no longer reproducible on the latest main. This is most likely due to the merge of https://github.com/Expensify/App/issues/33680, as expected.

Summary:
1. Original  [accepted proposal](https://github.com/Expensify/App/issues/34829#issuecomment-1904919422) identified the missing of Electron.Visibility() API's and defaulting to focus. 
2. After Implementation and thorough testing, an [undesired side effect](https://github.com/Expensify/App/pull/35004#issuecomment-1907218017) was identified. This was due to https://github.com/electron/electron/pull/41314. Great efforts from @blimpich to get it prioritized initially upstream, but it had to be reverted due to Linux issues.
3. Due to uncertainty in electron, I identified an alternate approach as mentioned in the bottom [here](https://github.com/Expensify/App/pull/35004#issuecomment-1907218017). I also realized it fixes a bigger picture issue with how we handle visibility and started referencing multiple tickets it would address. However a parallel initiative was occuring in https://github.com/Expensify/App/issues/33680 that also tackled this, and got merged in before.


Combing through tickets that dependent on this, I've retested and confirmed fixed state for  https://github.com/Expensify/App/issues/34698
and. https://github.com/Expensify/App/issues/35085 and provided an updated status for them.

If you can confirm this is also no longer reproducable @ishpaul777 I think we are good here.

**Comment by blimpich** [comment]
Thank you for your thorough summary @jeremy-croff!

**Comment by jeremy-croff** [comment]
@ishpaul777 I am perplexed.

https://github.com/Expensify/App/assets/157416545/8c9e7070-d775-4757-b5c0-602f5dc86e72

steps:
1.checkout main
2. pull latest
3. npm run web
4. npm run desktop
then video

**Comment by jeremy-croff** [comment]
I am able to reproduce, I have to have electron already focused to do so! Good catch.
I will try to get a good solution by today.

**Comment by jeremy-croff** [comment]
@ishpaul777 I am unable to come up with a decent solution at this point due to the occlusion issues; maybe it's time to consider opening it up again for new proposals.
I think what's going to make the most sense here is fix it upstream and make proposals for that.

**Comment by garrettmknight** [comment]
Still awaiting proposals.

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Chat isn't marked as read if user clicks on notification from Notification Center

### What is the root cause of that problem?

The macOS notification center remains open even after clicking on a notification, causing `browserWindow.isFocused()` to become false.

### What changes do you think we should make in order to solve the problem?

It's not possible to programmatically close or dismiss the macOS Notification Center from an Electron app.

We can trigger `readNewestAction`, each time we click the notification item. Change [this code](https://github.com/Expensify/App/blob/0eb9ef233ba5222032cb4865f9237160ddfbc92f/src/libs/actions/Report.ts#L2187-L2195) to:

```
        Modal.close(() => {
            const policyID = lastVisitedPath && extractPolicyIDFromPath(lastVisitedPath);
            const policyEmployeeAccountIDs = policyID ? getPolicyEmployeeAccountIDs(policyID) : [];
            const reportBelongsToWorkspace = policyID ? doesReportBelongToWorkspace(report, policyEmployeeAccountIDs, policyID) : false;
            if (!reportBelongsToWorkspace) {
                Navigation.navigateWithSwitchPolicyID({route: ROUTES.HOME});
            }
            Navigation.navigate(ROUTES.REPORT_WITH_ID.getRoute(reportID));
            readNewestAction(reportID);
        });

```

### What alternative solutions did you explore? (Optional)

**Comment by wildan-m** [comment]
Proposal [Updated](https://github.com/Expensify/App/issues/34829#issuecomment-2062877532)

Suggest creating `debouncedReadNewestAction` for further optimization.

**Comment by garrettmknight** [comment]
@ishpaul777 bump - if you want I can reassign while you're at less than 100%.

**Comment by wildan-m** [comment]
@ishpaul777 thanks for your feedback. My proposal [Updated](https://github.com/Expensify/App/issues/34829#issuecomment-2062877532)

Move prev main solution to alternative, Add new solution with new referrer querystring param

**Comment by wildan-m** [comment]
@ishpaul777 tested it on the web, I didn't notice any difference in behavior. However, since the issue only arises on desktops, it would be a good idea to limit the change to desktops for safety reasons.

**Comment by wildan-m** [comment]
> i say this becuase it doesn't directly fix root cause

I agree with it. It's a workaround, but we don't have many options since the `IsVisible` in the browser and electron (desktop) is expected to behave differently.

In Electron, the document.visibilityState inside isVisible function might behave differently because it's not a traditional web environment. The visibility of the Electron window might not be tied to the visibility of the tab in the same way it is in a browser. For example, an Electron app doesn't have tabs, so the concept of a tab being hidden doesn't really apply.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by garrettmknight** [comment]
@ishpaul777 any thoughts on @wildan-m's response? Separately, it looks like this didn't show up in tests. Is the fix still necessary for another reason?

**Comment by wildan-m** [comment]
@ishpaul777 [The PR](https://github.com/Expensify/App/pull/41484) is ready for review

**Comment by garrettmknight** [comment]
On hold for merge freeze, but we're ready to go otherwise.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by garrettmknight** [comment]
Payment Summary:
- Contributor: @wildan-m $500 paid via upwork
- Reviewer: @ishpaul777 $500 paid via upwork

@ishpaul777 can you complete the checklist?

**Comment by garrettmknight** [comment]
Nice one - done!
